migration/idempotent/017-metric-rollup.sql

```diff
@@ -0,0 +1,290 @@
+-- and calls the respective creation function. It returns true if metric rollup was created.
+CREATE OR REPLACE FUNCTION _prom_catalog.create_metric_rollup_view(rollup_schema TEXT, metric_name TEXT, table_name TEXT, resolution INTERVAL)
+RETURNS BOOLEAN AS
+$$
+    DECLARE
+        metric_type TEXT;
+
+    BEGIN
+        EXECUTE FORMAT('SELECT type FROM _prom_catalog.metadata WHERE metric_family = %L', metric_name) INTO metric_type;
+        IF metric_type IS NULL THEN
+            RAISE DEBUG '[Rollup] Skipping creation of metric rollup for %. REASON: metric_type not found', metric_name;
+            RETURN FALSE;
+        END IF;
+
+        CASE
+            WHEN metric_type = 'GAUGE' THEN
+                CALL _prom_catalog.create_rollup_for_gauge(rollup_schema, table_name, resolution);
+            WHEN metric_type = 'COUNTER' OR metric_type = 'HISTOGRAM' THEN
+                CALL _prom_catalog.create_rollup_for_counter(rollup_schema, table_name, resolution);
+            WHEN metric_type = 'SUMMARY' THEN
+                CALL _prom_catalog.create_rollup_for_summary(rollup_schema, table_name, resolution);
+            ELSE
+                RAISE WARNING '[Rollup] Skipping creation of metric rollup for %. REASON: invalid metric_type. Wanted {GAUGE, COUNTER, HISTOGRAM, SUMMARY}, received %', metric_name, metric_type;
+            END CASE;
+        RETURN TRUE;
+    END;
+$$
+LANGUAGE PLPGSQL;
+
+-- TODO: Temporary utilities for creation of metric rollups. These MUST be removed once we have SQL aggregate in Rust,
+-- since their behaviour is unreliable.
+CREATE FUNCTION _prom_catalog.counter_reset_sum(v DOUBLE PRECISION[]) RETURNS DOUBLE PRECISION AS
+$$
+DECLARE
+    reset_sum DOUBLE PRECISION := 0;
+    length INTEGER := cardinality(v);
+    i INTEGER := 1;
+    previous DOUBLE PRECISION;
+
+BEGIN
+    IF length < 2 THEN
+        RETURN 0;
+    END IF;
+    previous := v[1];
+    FOR i IN 2..length LOOP
+        IF v[i] < previous THEN
+            reset_sum := reset_sum + v[i];
+        END IF;
+    END LOOP;
+    RETURN reset_sum;
+END;
+$$
+LANGUAGE PLPGSQL IMMUTABLE;
+
+CREATE FUNCTION _prom_catalog.irate(v DOUBLE PRECISION[])
+RETURNS DOUBLE PRECISION AS
+$$
+    DECLARE
+        length INTEGER := cardinality(v);
+
+    BEGIN
+        IF length < 2 THEN
+            RETURN 0;
+        END IF;
+        return v[length] - v[length-1];
+    END;
+$$
+LANGUAGE PLPGSQL IMMUTABLE;
+
+-- Metric rollup creation by metric type
+CREATE OR REPLACE PROCEDURE _prom_catalog.create_rollup_for_gauge(rollup_schema TEXT, table_name TEXT, resolution INTERVAL)
+AS
+$$
+BEGIN
+    EXECUTE FORMAT(
+        'CREATE MATERIALIZED VIEW %1$I.%2$I WITH (timescaledb.continuous, timescaledb.materialized_only=true) AS
+            SELECT
+                timezone(
+                    %3$L,
+                    time_bucket(%4$L, time) AT TIME ZONE %3$L + %4$L
+                ) as time,
```

Comment:
can you explain what is happening here with timezones? I'm not following what's supposed to happen

Reply:
Sure, I have tried to keep the semantics similar to what's done [here](https://docs.timescale.com/promscale/latest/downsample-data/caggs/#example-of-a-promscale-continuous-aggregate) (See the note).

Reply:
okay, here's what i tested:

```

create table x (i int, t timestamptz);

insert into x (i, t) 
select 
    x,
    'today'::timestamptz + (interval '1 minute' * x) 
from generate_series(0, 60 * 24) x
;

select 
    time_bucket('1 hour'::interval, t) as hr,
    count(i) as nbr
from x
group by 1
order by 1
;


-- does time_bucket return a timestamptz when passed a timestamptz?
select 
    pg_typeof(time_bucket('1 hour'::interval, t)) as typeof,
    count(i) as nbr
from x
group by 1
order by 1
;
/*
┌──────────────────────────┬──────┐
│          typeof          │ nbr  │
├──────────────────────────┼──────┤
│ timestamp with time zone │ 1441 │
└──────────────────────────┴──────┘
*/


-- is it still a timestamptz if we pass in a timezone?
select 
    pg_typeof(time_bucket('1 hour'::interval, t, 'UTC')) as typeof,
    count(i) as nbr
from x
group by 1
order by 1
;
/*
┌──────────────────────────┬──────┐
│          typeof          │ nbr  │
├──────────────────────────┼──────┤
│ timestamp with time zone │ 1441 │
└──────────────────────────┴──────┘
*/
```

Reply:
so, if you pass in a `timestamptz` you will be returned a `timestamptz`. No need to add the `AT TIME ZONE` or wrap in the `timezone(...)` function

Reply:
it will by default be rendered in the `show timezone` timezone

Reply:
additionally, at the moment you are hardcoding the `timezone` parameter to the `time_bucket` function to `'UTC'`

- this is the default, so there's no need to pass it
- should we hardcode it to `'UTC'`, or should we use the database's `timezone` setting? does it matter? I'm not sure

Reply:
Thank you for looking into this.

I think all the `AT TIME ZONE` and wrapping with `timezone()` was done basically to ensure that we can correctly add the `+'1 hour'` (done [here](https://docs.timescale.com/promscale/latest/downsample-data/caggs/#example-of-a-promscale-continuous-aggregate)) for this reason `This is done to match the PromQL semantics of representing a bucket with the timestamp at the end of the bucket instead of the start of the bucket.`.

Now, from the above analysis, I think we can simply change
```postgresql
timezone('UTC', 
          time_bucket('1 hour', time) AT TIME ZONE 'UTC' +'1 hour')  
            as time
```
to
```postgresql
time_bucket('1 hour', time) + '1 hour' as time
```

Am I understanding this right?

Reply:
Okay, so when I made the change and again tested the rollup creation, I got this error

```
```

Reply:
Maybe thats all the reason for having the `timezone()` thing, @antekresic if you can take a look and share the idea behind using timezone [here](https://docs.timescale.com/promscale/latest/downsample-data/caggs/#example-of-a-promscale-continuous-aggregate).

Reply:
that's weird. `time_bucket` IS immutable. I wonder if `+` is not immutable...
<img width="612" alt="image" src="https://user-images.githubusercontent.com/4974668/198678302-8c37612b-8e46-4447-8330-aef40a147cdf.png">

Reply:
It's the addition :(
<img width="675" alt="image" src="https://user-images.githubusercontent.com/4974668/198678574-ff39b26a-c7d7-4c75-a646-c69e3a5150ac.png">

Reply:
https://www.postgresql.org/message-id/CANWftzJNW1hUZmRN9=tWn=K=bQrUhAOxkcogVNC72STp5-_czQ@mail.gmail.com

Reply:
In light of this, I think the original pattern (though unfortunate) makes sense. Sorry to sidetrack you for a bit.

Reply:
Not at all, this was an interesting investigation. But happy that we understood the root cause. Thank you!